File: ui/src/utils/IOId.js
// This file has a mirror in Kotlin server code.

/**
 * 
 * @param {String} step 
 * @param {String} nodeId 
 * @param {String} inputOrOutput Optional, id of input or output in the yaml. Null for pipeline inputs.
 * @returns the IO id: step@nodeId|inputOrOutput
 */
export function toIOId(step, nodeId, inputOrOutput) {
    return step + '@' + nodeId + (inputOrOutput ? '|' + inputOrOutput : '')
}

/**
 * @returns The path of the script that parses inputs / provides output.
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns script.yml
 * 
 * from script.yml@31|output
 * returns script.yml
 */
export function getScript(ioId) {
    const str = ioId.substring(0, ioId.lastIndexOf('@')) // pipeline@12|pipeline@23|script
    const lastPipe = str.lastIndexOf('|')
    return lastPipe === -1 ? str : str.substring(lastPipe + 1) // script
}

/**
 * @returns The unique identification of this step in the outer pipeline.
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json@12
 * 
 * from script.yml@31
 * returns script.yml@31
 */
export function getStepId(ioId) {
    const firstPipe = ioId.indexOf('|')
    return ioId.substring(
        0,
        firstPipe === -1 ? ioId.length : firstPipe
    )
}

/**
 * @returns Breadcrumbs from the outer pipeline to the script
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json@12|pipeline.json@23|script.yml@31
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31
 * returns pipeline.json@12|pipeline.json@23|script.yml@31
 * 
 * from script.yml@31|output
 * returns script.yml@31
 * 
 * from script.yml@31
 * returns script.yml@31
 */
export function getBreadcrumbs(ioId) {
    const lastPipe = ioId.lastIndexOf('|')
    const lastAt = ioId.lastIndexOf('@')
    if(lastAt < lastPipe) {
        return ioId.substring(0, lastPipe)
    }

    return ioId
}

/**
 * @returns The description file for this step in the outer pipeline.
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json
 */
export function getStepFile(ioId) {
    return ioId.substring(0, ioId.indexOf('@'))
}


/**
 * @returns The id of the react-flow node corresponding to this step, in the outer pipeline.
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns 12
 * 
 * from pipeline@12
 * returns 12
 */
export function getStepNodeId(ioId) {
    const firstPipe = ioId.indexOf('|')
    return ioId.substring(
        ioId.indexOf('@') + 1,
        firstPipe === -1 ? ioId.length : firstPipe
    )
}

/**
 * @returns The name of the output of this step.
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json@23|script.yml@31|output
 */
export function getStepOutput(ioId) {
    return ioId.substring(ioId.indexOf('|') + 1)
}

/**
 * @returns Output of the script (as defined in the script's yml)
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns output
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31
 * returns ""
 * 
 * from script.yml@31|output
 * returns output
 * 
 * from script.yml@31
 * returns ""
 */
export function getScriptOutput(ioId) {
    const lastPipe = ioId.lastIndexOf('|')
    const lastAt = ioId.lastIndexOf('@')
    if(lastAt < lastPipe) {
        return ioId.substring(lastPipe + 1)
    }

    return ""
}

/**
 * @returns The name of the input of this step.
 * 
 * from pipeline.json@12|pipeline.json@23|script.yml@31|input
 * returns pipeline.json@23|script.yml@31|input
 */
export function getStepInput(ioId) {
    return ioId.substring(ioId.indexOf('|') + 1)
}

/**
 * @returns The io id formatted in a human-readable way
 * 
 * from pipeline.json@12|pipeline.json@23|folder>script.yml@31|output
 * returns pipeline | pipeline | folder > script | output
 */
export function toDisplayString(ioId) {
    return ioId.replaceAll(/(.json|.yml)@\d+/g, '').replaceAll('>', ' > ').replaceAll('|', ' | ')
}

File: script-server/src/main/kotlin/org/geobon/pipeline/IOId.kt
package org.geobon.pipeline

/**
 * Id for an input or outputString
 * @param {String} step
 * @param {String} nodeId
 * @param {String} inputOrOutput Optional, id of input or output in the yaml. Null for pipeline inputs.
 */
data class IOId(val step: StepId, val inputOrOutput: String? = null) {

    /**
     * Get an IOId for a pipeline, from the pipeline's ID and the IOId of the step.
     */
    constructor(pipelineId: StepId, stepIoId:IOId) : this(pipelineId, stepIoId.toString())

    override fun toString(): String {
        return "${step.toString()}" + (if (inputOrOutput == null) "" else "|$inputOrOutput")
    }

    fun toBreadcrumbs(): String {
        return "${step.toBreadcrumbs()}" + (if (inputOrOutput == null) "" else "|$inputOrOutput")
    }
}

data class StepId(val step: String, val nodeId: String, val parent:StepId? = null) {
    override fun toString(): String {
        return 
            if(parent == null && step.isEmpty() && nodeId.isEmpty()) "" // Special case for root pipeline
            else "$step@$nodeId"
    }

    fun toBreadcrumbs(): String {
        return 
            if(parent == null) toString() 
            else "${parent.toBreadcrumbs()}|${toString()}"
    }
}

// -- The util functions below have a mirror in JS UI code.

/**
 * @returns The path of the script that parses inputs / provides output.
 *
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns script.yml
 *
 * from script.yml@31|output
 * returns script.yml
 */
fun getScript(ioId: String): String {
    val str = ioId.substring(0, ioId.lastIndexOf('@')) // pipeline@12|pipeline@23|script
    val lastPipe = str.lastIndexOf('|')
    return if (lastPipe == -1) str else str.substring(lastPipe + 1) // script
}

/**
 * @returns The unique identification of this step in the outer pipeline.
 *
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json@12
 *
 * from script.yml@31
 * returns script.yml@31
 */
fun getStepId(ioId: String): String {
    val firstPipe = ioId.indexOf('|')
    return ioId.substring(
        0,
        if (firstPipe == -1) ioId.length else firstPipe
    )
}

/**
 * @returns The description file for this step in the outer pipeline.
 *
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json
 */
fun getStepFile(ioId: String): String {
    return ioId.substring(0, ioId.indexOf('@'))
}


/**
 * @returns The id of the react-flow node corresponding to this step, in the outer pipeline.
 *
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns 12
 *
 * from pipeline@12
 * returns 12
 */
fun getStepNodeId(ioId: String): String {
    val firstPipe = ioId.indexOf('|')
    return ioId.substring(
        ioId.indexOf('@') + 1,
        if (firstPipe == -1) ioId.length else firstPipe
    )
}

/**
 * @returns The name of the output of this step.
 *
 * from pipeline.json@12|pipeline.json@23|script.yml@31|output
 * returns pipeline.json@23|script.yml@31|output
 */
fun getStepOutput(ioId: String): String? {
    val pipeIx = ioId.indexOf('|')
    if (pipeIx == -1)
        return null

    return ioId.substring(pipeIx + 1)
}

/**
 * @returns The name of the input of this step.
 *
 * from pipeline.json@12|pipeline.json@23|script.yml@31|input
 * returns pipeline.json@23|script.yml@31|input
 */
fun getStepInput(ioId: String): String? {
    return getStepOutput(ioId) // same naming convention as outputs
}